vyper/exceptions.py

```diff
@@ -51,74 +51,117 @@ def __str__(self):
 
 class PythonSyntaxException(ParserException):
     """
-    Conversion from error using ast.parse()
+    Invalid syntax.
+
+    This error is converted from errors raised during ast.parse
     """
     def __init__(self, syntax_error: SyntaxError, source_code: str):
         item = types.SimpleNamespace()  # TODO: Create an actual object for this
         item.lineno = syntax_error.lineno
         item.col_offset = syntax_error.offset
-        item.source_code = source_code
+        item.full_source_code = source_code
         super().__init__(message=f'SyntaxError: {syntax_error.msg}', item=item)
 
 
-class VariableDeclarationException(ParserException):
-    pass
+class SyntaxException(ParserException):
+    """Valid Python syntax, but invalid Vyper syntax."""
 
 
 class StructureException(ParserException):
-    pass
-
-
-class ConstancyViolationException(ParserException):
-    pass
-
-
-class NonPayableViolationException(ParserException):
-    pass
+    """TODO"""
+    # unknown / duplicate interface
+    # invalid map declaration
+    # invalid decorator
+    # use of semi-colon
+    # raise without reason
+    # iteration over nested list
+    # tuple length mismatch
+    # invalid assignment with/without value
+    # improperly formatted struct declaration
+    # invalid global keyword
+    # empty list/tuple
+    # unsupported statement type
+    # altering list during iteration
+    # invalid use of range()
 
 
 class InvalidLiteralException(ParserException):
-    pass
+    """Invalid literal value."""
+    # number out of range
+    # too many decimals
+    # invalid hexstring length
+    # comparison against None
 
 
 class InvalidTypeException(ParserException):
-    pass
+    """Invalid type declaration."""
+    # unknown type
+    # unknown unit
+    # declaring a mapping inside a function
+    # len() on a type without length
+    # decimal exponentiation
+    # no type given
 
 
-class TypeMismatchException(ParserException):
-    pass
+class VariableDeclarationException(ParserException):
+    """Invalid variable declaration."""
+    # invalid name
+    # already declared / shadows existing value
```

Comment:
already declared / undeclared are common themes across several exception types and so good candidates for new exceptions.

for shadowing / already declared vars: `NamespaceCollsion`
for references to things that don't exist: `UndeclaredDefinition`

Reply:
I'm good with this